ADR-0011: Agent WebSocket transport (#50582)
fleetdm/fleet: docs/Contributing/adr/0011-agent-websocket-transport.md

```mermaid
sequenceDiagram
    participant Agent
    participant Server

    loop every 10s
        Agent->>Server: anything for me?
        Server-->>Agent: no (99.7%)
    end

    Note over Server: live query created
    loop next 10s tick
        Agent->>Server: anything for me?
        Server-->>Agent: yes, here's your query
    end
```

### How it works with WebSockets (push)

```mermaid
sequenceDiagram
    participant Agent
    participant Server

    Agent->>Server: poll config (existing mechanism)
    Server-->>Agent: config includes websocket_enabled=true
    Agent->>Server: open WebSocket (with jitter delay)
    Note over Agent,Server: silent until needed

    Note over Server: live query created
    Server->>Agent: check now (WebSocket push)
    Agent->>Server: distributed/read (HTTP)
    Server-->>Agent: here's your query
```

### Who decides and who connects

The **server** decides whether WebSocket transport is active, not the agent. When the feature flag is enabled on the server, it includes a WebSocket directive in the agent's configuration response (delivered through the existing config polling mechanism). On the next config poll:

> - **New agent (supports WebSockets):** reads the directive and opens a WebSocket connection to the server. To avoid a thundering herd when the feature is first enabled, each agent applies a random jitter delay before connecting.

Reviewer: I don't think a random jitter help here because the connection attempts of clients is also random. With this new websockets approach, the actual thundering herd protection can be implemented in Fleet now (e.g. after several minutes of downtime, instead of sending all agents "we have new queries for you now" we can do that progressively)

- **Why not chosen:** WebSockets support multiplexing multiple notification types on a single connection, which is critical for future phases. Long polling forfeits this and adds connection churn.

### Server-Sent Events (SSE)

The server pushes events to the agent over a long-lived HTTP response using the `text/event-stream` content type. The agent opens one GET request and the server streams events as they occur.

- **Pros:** Simpler than WebSockets. Built on standard HTTP, works through most proxies. Native reconnection with `Last-Event-ID`. One connection can carry multiple event types.
- **Cons:** Unidirectional (server to agent only). The agent cannot send data back over the same connection, so all agent-to-server communication still requires separate HTTP calls (which is also true of our WebSocket design). More critically, SSE requires HTTP/1.1 chunked transfer or HTTP/2. ALB-to-target communication in Fleet's topology is HTTP/1.1, which limits concurrent SSE streams per browser/client. Enterprise middleboxes (TLS-inspecting proxies) often buffer chunked responses, breaking the real-time delivery that SSE depends on.

> - **Why not chosen:** The middlebox buffering problem is the same class of issue that led Kolide to add an HTTP fallback for gRPC. WebSockets have a cleaner upgrade mechanism that middleboxes handle better in practice. Fleet already runs WebSockets in production (live query results in the UI), so operational experience exists.

Reviewer 2: We also already do SSE, and we solved the middlebox buffering issue (which we had with ngrok):
```
ue.GET("/api/_version_/fleet/android_enterprise/signup_sse", enterpriseSSE, nil)
```

SSE Pro is that it doesn't require an HTTP/1.1 Upgrade handshake that every hop must honor, so it is more likely to work through corporate proxies.

Author: @getvictor, Lucas is DRI for this ADR starting today, so for any comments, please @-mention him to make sure he sees them.

cc @lucasmrod

Reviewer 3: Suggesting a change to clarify the reasoning in SSE vs. WebSockets.

Reviewer: See https://github.com/fleetdm/fleet/pull/50582/changes/da89395ca9403aae58c66aabedd06d83984ce2f9.

> - **Cons:** Unidirectional (server to agent only). The agent cannot send data back over the same connection, so all agent-to-server communication still requires separate HTTP calls (which is also true of our WebSocket design). More critically, SSE requires HTTP/1.1 chunked transfer or HTTP/2. ALB-to-target communication in Fleet's topology is HTTP/1.1, which limits concurrent SSE streams per browser/client. Enterprise middleboxes (TLS-inspecting proxies) often buffer chunked responses, breaking the real-time delivery that SSE depends on.
> - **Why not chosen:** The middlebox buffering problem is the same class of issue that led Kolide to add an HTTP fallback for gRPC. WebSockets have a cleaner upgrade mechanism that middleboxes handle better in practice. Fleet already runs WebSockets in production (live query results in the UI), so operational experience exists.

Reviewer 3: ```suggestion
- **Con: no liveness detection.** SSE is unidirectional (server to agent only). There is no pong, so the server cannot tell a live peer from a half-open connection (laptop asleep, NAT entry dropped) until a TCP write fails. This design depends on accurate per-instance connection accounting (the interval check job, rebalancing counts, connection caps, closing revoked node keys), and dead peers pollute all of it, with nudges silently lost into dead sockets.
- **Con: silent failure through middleboxes.** Buffering middleboxes (TLS-inspecting proxies) break SSE without breaking the connection: the stream establishes, but events arrive late or never, and the agent cannot distinguish "no work" from "nudges stuck in a proxy buffer" because the protocol's steady state is silence. WebSocket failures are loud (the upgrade handshake is rejected), which triggers the polling fallback deterministically.
- **Why not chosen:** Fleet does run SSE in production (the Android enterprise signup stream), but that is a single short-lived browser connection, and its middlebox fix (`X-Accel-Buffering: no` plus heartbeat comments) is a hint that cooperative reverse proxies honor, not a guarantee enterprise gear does. It is true that SSE avoids the Upgrade handshake and may traverse some corporate proxies that block WebSockets, but on those networks the WebSocket design degrades cleanly to polling, which is the correct outcome. The deciding factors are bidirectional liveness (ping/pong) and detectable failure, both of which the connection-accounting design requires and SSE would force us to rebuild by hand.
```

# ADR-0011: Agent WebSocket transport

## Status

> Proposed

Reviewer 3: ```suggestion
Approved
```

## Date

> 2026-08-04

Reviewer 3: ```suggestion
2026-08-13
```

# ADR-0011: Agent WebSocket Transport

## Status

Proposed

## Authors

- @lucasmrod
- @lukeheath
- @sharon-fdm

## Date

2026-08-04

## Table of Contents

1. [What & Why](#1-what--why)
2. [How Orbit Proxies osquery](#2-how-orbit-proxies-osquery)
3. [Security Analysis](#3-security-analysis)
4. [Deployment & Rollout](#4-deployment--rollout)
5. [Consequences](#consequences)
6. [Alternatives Considered](#alternatives-considered)
7. [References](#references)

---

## 1. What & Why

### The problem

> Today, every Fleet agent polls the server on fixed timers. The biggest offender is `distributed/read` (live query check-in): every host asks "anything for me?" every 10 seconds, and 99.7% of the time the answer is "no."

Reviewer: ```suggestion
Today, every Fleet agent polls the server on fixed timers. The biggest offender is `distributed/read` (query check-in): every host asks "anything for me?" every 10 seconds, and 99.7% of the time the answer is "no."
```

- **Old agent (no WebSocket support):** ignores the unknown directive and continues polling as before. No harm done.

> **The server is always in control.** Disabling the feature flag immediately stops new WebSocket connections on the next config cycle. Every agent falls back to polling. No agent action required, no rollback needed, no downtime.

### What travels over `distributed/read` today

The `distributed/read` endpoint is not just for live queries. Three distinct features share it, each with its own server-side interval:

| Feature | How it gets to the agent | Server includes it when... |
|---|---|---|
| **Live queries** | `distributed/read` | A campaign targets the host |
| **Policies** | `distributed/read` | PolicyUpdateInterval has elapsed (~1 hour) |
| **Software ingestion** | `distributed/read` (detail queries) | DetailUpdateInterval has elapsed (~1 hour) |

The agent polls `distributed/read` every 10 seconds. The server decides what to include based on these intervals. In steady state, most polls return empty because no live query is active and the hourly intervals have not elapsed.

> **Scheduled queries (reports)** use a different channel: they are delivered via `/api/osquery/config` as part of the osquery pack configuration. This is covered by the config endpoint, not `distributed/read`.

Reviewer: ```suggestion
> **Scheduled queries (reports)** use a different channel: they are delivered via `/api/osquery/config` as part of the osquery pack configuration.
```

### How each feature works today (polling)

> The diagrams below are simplified for illustration.

> **Live queries** (the primary target of this ADR):

Reviewer: `distributed/read` is used for policies, labels, host vitals (which include software), AND live queries, so the primary target cannot be just live queries, but all functionality that is delivered via distributed/read.
```suggestion
**Live queries**:
```

```mermaid
sequenceDiagram
    participant Agent
    participant Server
    loop every 10s
        Agent->>Server: distributed/read
        Server-->>Agent: empty (no active campaign)
    end
    Note over Server: admin runs a live query
    Agent->>Server: distributed/read (next 10s tick)
    Server-->>Agent: here is your query
    Agent->>Server: distributed/write (results)
```

**Policies** (delivered via distributed/read, but only every ~1 hour):

```mermaid
sequenceDiagram
    participant Agent
    participant Server
    loop every 10s
        Agent->>Server: distributed/read
        Server-->>Agent: empty (policy interval not elapsed)
    end
    Note over Server: 1 hour elapses
    Agent->>Server: distributed/read
    Server-->>Agent: policy queries
    Agent->>Server: distributed/write (pass/fail results)
```

> **Software ingestion** (delivered via distributed/read as detail queries, ~1 hour):

Reviewer: ```suggestion
**Host vitals ingestion** (which include software, delivered via distributed/read as detail queries, ~1 hour):
```

```mermaid
sequenceDiagram
    participant Agent
    participant Server
    loop every 10s
        Agent->>Server: distributed/read
        Server-->>Agent: empty (detail interval not elapsed)
    end
    Note over Server: 1 hour elapses
    Agent->>Server: distributed/read
    Server-->>Agent: software inventory queries
    Agent->>Server: distributed/write (software list)
```

> **Scheduled queries / reports** (different channel, via config):

Reviewer: This should be removed. This part won't change on this iteration.

```mermaid
sequenceDiagram
    participant Agent
    participant Server
    loop every 60s
        Agent->>Server: /api/osquery/config
        Server-->>Agent: config with pack/schedule (usually unchanged)
    end
    Note over Agent: osquery runs scheduled queries locally on their own timers
    Agent->>Server: /api/osquery/log (result logs)
```

### The WebSocket is a notification channel only

> The WebSocket does **not** replace any existing functionality. It acts purely as a notification channel: the server sends a short "check now" signal, and the agent then performs the same HTTP calls it always has. No query data, no config payloads, no results travel over the WebSocket. Everything that works today continues to work exactly the same way. The only difference is that the agent no longer asks on a blind timer; it asks when told to.

Reviewer: The "check now" signal should have a `type`, which in this Phase 1 will always be `type=distributed/read`, in the future, type could be `type=orbit/config`, etc. etc.

> Because live queries, policies, and software ingestion all share `distributed/read`, a single WebSocket nudge type covers all three. The agent does not need to know which feature triggered the nudge; it just calls `distributed/read` and the server returns whatever is due.

Reviewer: ```suggestion
Because live queries, policies, labels, and host vitals (e.g. software) ingestion all share `distributed/read`, a single WebSocket nudge type covers all three. The agent does not need to know which feature triggered the nudge; it just calls `distributed/read` and the server returns whatever is due.
```

> **Phase 1 (POC):** The notification channel covers `distributed/read` only (live queries, policies, software ingestion). This is the biggest offender (38.2% of all traffic, 99.7% empty) and proves the mechanism end to end.

Reviewer: ```suggestion
**Phase 1 (POC):** The notification channel covers `distributed/read` only (live queries, labels, policies, host vitals). This is the biggest offender (38.2% of all traffic, 99.7% empty) and proves the mechanism end to end.
```

**Future phases:** The same WebSocket carries notifications for additional channels:

| Phase | Channel | Current polling | What the nudge means |
|---|---|---|---|
| 1 (POC) | `distributed/read` | every 10s | "there is work for you" (live queries, policies, or software ingestion) |
| Future | orbit config | every 30s | "your config changed" |
| Future | Fleet Desktop check-in | every 5m | "there is something to show the user" |
| Future | osquery config | every 60s | "your osquery config changed" |

Each new channel is a new message type on the same connection. The pattern is always the same: the server sends a short nudge, the agent performs the corresponding HTTP call. One socket, many notification types, added incrementally.

### Keepalive and agent liveness

> The server sends a WebSocket ping every **5 minutes**. This serves two purposes: confirming the agent is still alive, and preventing the load balancer (e.g. AWS ALB) from killing the connection due to idle timeout. The ALB idle timeout should be configured to a value longer than the ping interval (e.g. 10 minutes). If a pong is not received within 30 seconds, the server considers the connection dead and closes it. The agent's normal reconnection logic (with jitter) handles recovery.

Reviewer: In reality we might not be able to set the ping interval as high as 5 minutes. Worth checking with infrastructure (my guess is that proxies, NATs, or other middle layers will drop unused connections sooner than 5 minutes)

@rfairburn

### Feature flag and activation

> The WebSocket transport is controlled by a server-side command-line flag (e.g. `--enable_websocket_transport`). It is not exposed in the UI, not documented, and not available through any API or GitOps configuration. The only way to enable it is to start the Fleet server with this flag. When disabled (the default), the directive is absent and all agents poll as usual.

Reviewer: Let's also add the requirement in fleetd that if the websocket cannot be established then the agent will resort to the old protocol.

### Thundering herd mitigation

When the feature flag is first enabled, every agent receives the WebSocket directive on its next config poll. If all agents immediately open a WebSocket, the server receives tens of thousands of connection attempts within seconds. The same problem occurs when the server restarts: every held connection drops, and all agents try to reconnect at once.

> To prevent this, each agent applies a random jitter delay before connecting:

Reviewer: With this new approach the Fleet server has more power to control the thundering herd. (Can send "check now" signals progressively instead of sending "check now" to all.)

```mermaid
sequenceDiagram
    autonumber
    participant OSQ as osquery
    participant Orbit as orbit (distributed plugin)
    participant S2 as Fleet server B (holds this agent's WebSocket)
    participant S1 as Fleet server A (creates campaign)
    participant Redis

    Note over OSQ,S1: SETUP
    Orbit->>S2: poll config (existing HTTP)
    S2-->>Orbit: config with websocket_enabled=true
    Orbit->>Orbit: register as osquery's distributed plugin
    Orbit->>Orbit: flip --distributed_plugin to local extension
    Orbit->>S2: open WebSocket (with jitter delay)
    S2->>S2: hold connection

    Note over OSQ,Orbit: STEADY STATE (nothing to do)
    loop every 10s
        OSQ->>Orbit: any queries? (localhost thrift)
        Orbit-->>OSQ: no (answered from memory)
    end
    Note over OSQ,S2: zero network traffic

    Note over S1,Redis: LIVE QUERY CREATED (on server A)
    S1->>Redis: write targeting state + PUBLISH wake-up (new)
    Redis-->>S2: server B receives wake-up (new)
    S2->>Orbit: check now (over WebSocket)
    Orbit->>S2: distributed/read (HTTP)
    S2-->>Orbit: query payload

>     Orbit->>Orbit: cache query in memory

Reviewer: This diagram is for live queries only. It has to be updated with a cron that will check which hosts need a "check now" message (e.g. host 1 has stale policies, host 2 has stale labels, host 3 has stale host vitals, send "check now" to the three, etc.)

    Note over OSQ,Orbit: NEXT LOCAL POLL (within 10s)
    OSQ->>Orbit: any queries? (localhost thrift)
    Orbit-->>OSQ: yes, here is your query
    OSQ->>OSQ: execute query
    OSQ->>Orbit: results (localhost thrift)
    Orbit->>S2: distributed/write (HTTP)
```

> **Note on Redis pub/sub (new mechanism):** Today, campaign targeting is written to Redis keys and each server instance discovers it independently when a host polls. There is no inter-instance notification. With WebSockets, the server instance holding an agent's connection may not be the one that created the campaign. To solve this, campaign creation adds a Redis `PUBLISH` on a wake-up channel. All server instances subscribe to this channel and nudge the relevant agents they hold. Redis pub/sub already exists in Fleet for streaming query results back; this extends the same pattern for the wake-up signal.

Reviewer: See https://github.com/fleetdm/fleet/pull/50582/changes#r3738192400.

> **Orbit proxies all osquery traffic, not just distributed queries.** With orbit acting as osquery's proxy, all osquery-to-server communication flows through orbit. This includes:

Reviewer: This is a future phase, this first phase is ONLY distributed/read traffic.

**Key mitigation:** The WebSocket upgrade endpoint must authenticate the node key **before** allocating connection resources. A failed auth check should return `401` and close the TCP connection immediately, not hold it open.

### Attack surface comparison

The WebSocket does not introduce new data flows or new trust boundaries:

| Concern | Today (polling) | With WebSocket |
|---|---|---|
| Encryption | TLS (HTTPS) | TLS (WSS), same certs |
| Authentication | Orbit node key per request | Orbit node key at upgrade, then persistent |
| Data on the wire | Full query payloads, config, results | Only "check now" nudges (a few bytes); all payloads still go over HTTP |
| Server-to-agent channel | None (pull only) | Yes, but carries no sensitive data |
| Spoofing risk | Attacker needs valid node key | Same, node key required at upgrade |

The new server-to-agent channel (the nudge) carries no sensitive information. The worst an attacker could do if they compromised the WebSocket channel is send a false "check now" signal, which would cause the agent to make one extra HTTP `distributed/read` call. This is equivalent to what already happens every 10 seconds today.

### Redis pub/sub

> The wake-up signal is published through Redis pub/sub so all server instances can relay nudges to agents connected to them. Redis should be deployed on a private network with authentication enabled (Fleet's existing Redis configuration). The pub/sub message contains only the campaign ID and target host identifiers, not query content or results.

Reviewer: See https://github.com/fleetdm/fleet/pull/50582/changes#r3738192400.

### Proposed rollout order

**Stage 1: Dogfood.** Enable the feature flag on Fleet's internal Dogfood server. Validate stability, connection behavior, keepalive, reconnection, and fallback. This is a low-risk environment where we can observe the feature under real (but internal) usage.

**Stage 2: Volunteer customer.** Identify a customer willing to opt in early. Coordinate with Customer Success on whether to offer an incentive (e.g. reduced hosting cost) in exchange for being an early adopter. Run with the feature enabled, monitor closely, and gather feedback.

**Stage 3: Broader managed cloud rollout.** Expand to additional customers. Volunteers first, then progressively enable on more deployments as confidence grows. At each step, validate cost savings match expectations and no regressions occur.

**Stage 4: Document and publish.** Once the feature is proven at scale across managed cloud, document the feature flag and make it available to self-hosted customers who want to enable it on their own infrastructure.

> > **Open question for @lucasmrod and @lukeheath:** What do you think of this rollout order? Any concerns or suggestions?

Reviewer: Looks good to me.

---

## Consequences

### Where the cost savings come from

With ECS Fargate, we pay per container (vCPU + memory allocation), not per CPU cycle. Keeping 15 instances idle saves nothing on compute. The savings come from three areas:

1. **Fewer/smaller server instances.** With 1.2B fewer HTTP requests/day, the cluster needs far less compute capacity. The instance count can be reduced (e.g. 15 to 8) or instances can be downsized.
2. **Reduced ALB costs.** ALB pricing is based on LCUs (new connections, active connections, bandwidth). Eliminating most HTTP requests dramatically reduces LCU usage.
3. **Reduced data transfer.** Fewer HTTP responses means less egress.

| | Today | With WebSockets |
|---|---|---|
| Server instances | 15 | Fewer (right-sized to actual load) |
| HTTP requests/day | ~1.2B | Near zero in steady state |
| Infra cost/day | $122 | ~$52 (measured in load test) |
| What drives the savings | -- | Fewer instances + lower ALB/transfer costs |

### The tension: fewer instances vs. WebSocket headroom

> Reducing instance count increases WebSocket density per instance. Each instance must still handle HTTP bursts when nudges fire (agents do `distributed/read` and `distributed/write` over HTTP). The connection budget per instance must account for both.

Reviewer: Given this is a new mode of communication we will learn the instance count needed for host count as we deploy this progressively.

> **The server is always in control.** Disabling the feature flag immediately stops new WebSocket connections on the next config cycle. Every agent falls back to polling. No agent action required, no rollback needed, no downtime.

### What travels over `distributed/read` today

The `distributed/read` endpoint is not just for live queries. Three distinct features share it, each with its own server-side interval:

> | Feature | How it gets to the agent | Server includes it when... |

Reviewer: Label queries (LabelUpdateInterval) and vitals too (DetailUpdateInterval, which include software), see all queries in https://github.com/fleetdm/fleet/blob/main/docs/Contributing/product-groups/orchestration/understanding-host-vitals.md.
